lib/src/client.dart

```diff
@@ -51,42 +53,154 @@ class ChannelOptions {
+  /// made on this connection.
+  Future<Null> terminate() {
+    _isShutdown = true;
+    // TODO(jakobr): Manage streams, close them immediately.
+    return _transport.terminate();
+  }
+
+  ClientTransportStream sendRequest(
+      bool useTls, String authority, String path, CallOptions options) {
+    final headers = createCallHeaders(useTls, authority, path, options);
+    final stream = _transport.makeRequest(headers);
+    // TODO(jakobr): Manage streams. Subscribe to stream state changes.
+    return stream;
+  }
+}
+
+/// A channel to a virtual RPC endpoint.
+///
+/// For each RPC, the channel picks a [ClientConnection] to dispatch the call.
+/// RPCs on the same channel may be sent to different connections, depending on
+/// load balancing settings.
 class ClientChannel {
   final String host;
   final int port;
   final ChannelOptions options;
 
-  final List<Socket> _sockets = [];
-  final List<TransportConnection> _connections = [];
+  final _connections = <ClientConnection>[];
+
+  bool _isShutdown = false;
 
   ClientChannel(this.host,
       {this.port = 443, this.options = const ChannelOptions()});
 
+  String get authority => options.authority ?? host;
+
+  void _shutdownCheck([Function() cleanup]) {
+    if (!_isShutdown) return;
+    cleanup();
+    throw new GrpcError.unavailable('Channel shutting down.');
+  }
+
+  /// Shut down this channel.
+  ///
+  /// No further calls can be made on this channel. Calls already in progress
+  /// will be allowed to complete.
+  Future<Null> shutdown() {
+    _isShutdown = true;
+    return Future.wait(_connections.map((c) => c.shutdown()));
+  }
+
+  /// Terminate this channel.
+  ///
+  /// Calls already in progress will be terminated. No further calls can be made
+  /// on this channel.
+  Future<Null> terminate() {
+    _isShutdown = true;
+    return Future.wait(_connections.map((c) => c.terminate()));
+  }
+
   /// Returns a connection to this [Channel]'s RPC endpoint. The connection may
   /// be shared between multiple RPCs.
-  Future<ClientTransportConnection> connect() async {
+  Future<ClientConnection> connect() async {
+    _shutdownCheck();
     final securityContext = options.securityContext;
 
-    Socket socket;
-    if (securityContext == null) {
-      socket = await Socket.connect(host, port);
-    } else {
-      socket = await SecureSocket.connect(host, port,
-          context: securityContext, supportedProtocols: ['grpc-exp', 'h2']);
+    var socket = await Socket.connect(host, port);
+    _shutdownCheck(() => socket.destroy());
+    if (securityContext != null) {
+      socket = await SecureSocket.secure(socket,
+          host: authority, context: securityContext);
+      _shutdownCheck(() => socket.destroy());
     }
-    _sockets.add(socket);
-    final connection = new ClientTransportConnection.viaSocket(socket);
+    final connection =
+        new ClientConnection(new ClientTransportConnection.viaSocket(socket));
```

Comment:
Seems a bit weird that a `ClientConnection` is created from a `ClientTransportConnection` rather than a `ClientTransport`.